### ISSUE TITLE
i18n: Wrong locale on getServerSideProps when using next.config.js rewrite

### ISSUE THREAD

**Issue by jpedroribeiro** [vercel/next.js#53988]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: linux
      Arch: x64
      Version: #1 SMP PREEMPT_DYNAMIC Sun Aug  6 20:05:33 UTC 2023
    Binaries:
      Node: 18.14.2
      npm: 9.5.0
      Yarn: 1.22.19
      pnpm: N/A
    Relevant Packages:
      next: 13.4.14-canary.5
      eslint-config-next: 13.4.12
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.0.4
    Next.js Config:
      output: N/A
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

Data fetching (gS(S)P, getInitialProps), Internationalization (i18n), Routing (next/router, next/navigation, next/link)

### Link to the code that reproduces this issue or a replay of the bug

https://codesandbox.io/p/sandbox/routing-i18n-bug-gfc94t?file=%2Fnext.config.js%3A1%2C1

### To Reproduce

If you click on _"Link to Page1 [PT](https://gfc94t-3000.csb.app/pt/page1/)"_ you will see that both `props` and `router` will output the correct locale, `pt`. 

However, if you click on _"Link to Pagina1 (Redirect to Page1) [PT](https://gfc94t-3000.csb.app/pt/pagina1/)"_, `props` will be the `defaultLocale` (which is `en`), whereas `router` be the expected locale, `pt`.

The only difference between both links is that the second one works via a `next.config.js` `rewrite` property.

### Describe the Bug

`getServerSideProps` is not getting the correct `locale` if the route was using a `next.config.js` `rewrite`.

This was **working** as expected up until and including version `13.2.5-canary.16`. On version `13.2.5-canary.17` it has **stopped working** and has not being fixed yet (example on [CodeSandbox](https://codesandbox.io/p/sandbox/routing-i18n-bug-gfc94t?file=%2Fnext.config.js%3A1%2C1) is using the latest version `13.4.14-canary.5`).

### Expected Behavior

`getServerSideProps` should have the correct `locale` value regardless of the routing strategy.

**Comment by iljamulders** [comment]
Same as:

https://github.com/vercel/next.js/issues/51923
https://github.com/vercel/next.js/issues/49883
https://github.com/vercel/next.js/issues/51695

This has been a bug for 4+ months, and everyone that uses i18n and rewrites/redirects is at a standstill...
13.2.4 is the last working version.

**Comment by pmcflepage** [comment]
Huge thanks to @iljamulders I finally solve it returning to 13.2.4!

**Comment by przemekga** [comment]
same as https://github.com/vercel/next.js/issues/50498

**Comment by vinnycodes** [comment]
Thank you @iljamulders 
This is exactly what we needed.

**Comment by dejanstrancar** [comment]
@iljamulders thanks for the heads up. Looks like last working version for us is `next@13.4.12`.

**Comment by jpedroribeiro** [comment]
I've updated my example to use the latest version `13.5.1` and it seems to have fixed the issue. https://codesandbox.io/p/sandbox/routing-i18n-bug-gfc94t?file=%2Fpackage.json%3A18%2C20

**Comment by raRaRa** [comment]
Still broken for me in 13.5.1 if middleware is being used, even though the middleware is not even triggered for the paths. It just needs to exist. Ref: https://github.com/vercel/next.js/issues/54217

**Comment by jpedroribeiro** [comment]
@raRaRa I had to remove `matcher` from Middleware and use [Conditional Statement](https://nextjs.org/docs/pages/building-your-application/routing/middleware#conditional-statements) instead.

**Comment by raRaRa** [comment]
I see, that makes the middleware run for every URL even though it's not needed (more costly AFAIK). But either way this is still a bug.

**Comment by jpedroribeiro** [comment]
Yeah you're right @raRaRa, leaving out the `matcher` is not an optimal solution.

**Comment by albertodeago** [comment]
Have you tried on latest version? I think this is not happening anymore on next 14.1.0

**Comment by Stanley-Codelevate** [comment]
> Have you tried on latest version? I think this is not happening anymore on next 14.1.0

tried today, still broken

**Comment by Degn** [comment]
I still experience this in next 14.1.4.
Adding `"/(en|de|fr|da|es)/:path*"` in the middleware matcher worked for me. Not optimal but it works for now...
`en|de|fr|da|es` is my hardcoded locales.

**Comment by tcg** [comment]
I think I'm running into this on 14.2.4. Found this by searching. 

I have a link like `/fr/path/to/stuff` or `/de/path/to/stuff`. When I visit it directly, the `locale` in the getServerSideProps context is `fr` or `de`, respectively. 

But clicking a link on the page, whose destination is e.g., `href="/fr/path/to/otherstuff"`, the `locale` provided to `getServerSideProps` is the default locale (`en`). 

This seems to be the most relevant issue. So just pitching a vote here  😅

I just discovered this, so haven't yet had time to see if there are open PRs related to this.

**Comment by tcg** [comment]
I tried out https://github.com/vercel/next.js/pull/61561 with `yarn patch`, and it appears to resolve this issue. 

I see there are some test failures there, but I'm new to Next internals, so don't know if that's flake or legit.

**Comment by tcg** [comment]
> @raRaRa I had to remove `matcher` from Middleware and use [Conditional Statement](https://nextjs.org/docs/pages/building-your-application/routing/middleware#conditional-statements) instead.

Removing `matcher` (actually, the whole `config` export, which only contained matcher) appears to resolve this issue for me as well! 

Thankfully, we don't appear to require that. Our conditional check is pretty lightweight as-is.

**Comment by StokleyO** [comment]
> @raRaRa I had to remove `matcher` from Middleware and use [Conditional Statement](https://nextjs.org/docs/pages/building-your-application/routing/middleware#conditional-statements) instead.

do you have an example of what it looks like to replace a matcher with a conditional statement?

**Comment by StokleyO** [comment]
Solution I found: add the route paths that use getServerSideProps (where the locale mismatch happens) to your config matcher array in middleware.js. This solved the problem for me and allowed me to continue using my config matchers

**Comment by nhayhoc** [comment]
same here

**Comment by armanTakmazyan** [comment]
I’m experiencing the same issue. I have a **middleware** configured with a **matcher** for a specific **URL**. However, the **middleware** isn’t even triggered on the page where the **issue** occurs. **Removing** the **middleware** resolves the problem.

**Comment by marcoshernanz** [comment]
Thanks for reporting this. I can reproduce the historical regression in `13.2.5-canary.17`, but not in `16.3.1` or `16.3.1-canary.25`. The middleware matcher variant was fixed by #68947 and has an end-to-end regression test. Closing as resolved. If this still reproduces on a current release, please open a new issue with an updated minimal reproduction.
